ledger/participant-integration-api/src/main/scala/platform/store/backend/common/ContractStorageBackendTemplate.scala

```diff
@@ -114,6 +118,44 @@ class ContractStorageBackendTemplate(
            ORDER BY event_sequential_id DESC
            FETCH NEXT 1 ROW ONLY"""
       .as(fullDetailsContractRowParser.singleOpt)(connection)
+      .map(_._2)
+  }
+
+  override def contractStates(contractIds: Seq[ContractId], before: Offset)(
+      connection: Connection
+  ): Map[ContractId, ContractStorageBackend.RawContractState] = if (contractIds.isEmpty) Map.empty
+  else {
+    import com.daml.platform.store.backend.Conversions.ContractIdToStatement
+    import com.daml.platform.store.backend.Conversions.OffsetToStatement
+    SQL"""WITH joined_rows (contract_id, create_event, archive_event, template_id, create_witness, archive_witness,
```

Comment:
I was going to say that this query looks complicated. But the ideas of improvement don't necessary pan out simpler. For instance 
- a query that starts of with union of creates and archives -> then proceeds to group by contract_id is simple enough but there are no obvious criteria of selecting values for some of the columns
- a query that starts of with union of creates and archives -> then proceeds to partition by contract_id and rank by event type, gets a bit messy too.
Unless someone has a better idea I tend to say this is as good as it gets.

Reply:
To capture the results for the future:

This is the bulk query:
```
----
 Append  (cost=4.45..4.52 rows=2 width=122) (actual time=0.054..0.057 rows=1 loops=1)
   CTE joined_rows
     ->  Nested Loop Left Join  (cost=0.00..4.45 rows=1 width=625) (actual time=0.038..0.042 rows=1 loops=1)
           Join Filter: (c.contract_id = a.contract_id)
           ->  Index Scan using participant_events_create_contract_id_idx on participant_events_create c  (cost=0.00..2.22 rows=1 width=577) (actual time=0.022..0.023 rows=1 loops=1)
                 Index Cond: (contract_id = '004b4007da46aa40f0027345ba00bd542f581938b85fc584bb0be1e6c405847368ca011220116fd8c958d7dc0e268094b276b812c84fa41795892b78e2be94fd0cd303370e'::text)
                 Filter: (event_offset < '92862790'::text)
           ->  Index Scan using participant_events_consuming_exercise_contract_id_idx on participant_events_consuming_exercise a  (cost=0.00..2.22 rows=1 width=190) (actual time=0.010..0.012 rows=1 loops=1)
                 Index Cond: (contract_id = '004b4007da46aa40f0027345ba00bd542f581938b85fc584bb0be1e6c405847368ca011220116fd8c958d7dc0e268094b276b812c84fa41795892b78e2be94fd0cd303370e'::text)
                 Filter: (event_offset < '92862790'::text)
   ->  CTE Scan on joined_rows  (cost=0.00..0.02 rows=1 width=122) (actual time=0.048..0.048 rows=0 loops=1)
         Filter: (archive_event IS NULL)
         Rows Removed by Filter: 1
   ->  CTE Scan on joined_rows joined_rows_1  (cost=0.00..0.02 rows=1 width=122) (actual time=0.004..0.005 rows=1 loops=1)
         Filter: (archive_event IS NOT NULL)
 Planning Time: 0.664 ms
 Execution Time: 0.126 ms
(17 rows)
```

while this is the "single lookup":

```
 Limit  (cost=4.48..4.48 rows=1 width=407) (actual time=0.053..0.055 rows=1 loops=1)
   ->  Sort  (cost=4.48..4.49 rows=2 width=407) (actual time=0.052..0.053 rows=1 loops=1)
         Sort Key: participant_events_create.event_sequential_id DESC
         Sort Method: quicksort  Memory: 26kB
         ->  Append  (cost=0.00..4.47 rows=2 width=406) (actual time=0.023..0.041 rows=2 loops=1)
               ->  Index Scan using participant_events_create_contract_id_idx on participant_events_create  (cost=0.00..2.22 rows=1 width=581) (actual time=0.022..0.023 rows=1 loops=1)
                     Index Cond: (contract_id = '004b4007da46aa40f0027345ba00bd542f581938b85fc584bb0be1e6c405847368ca011220116fd8c958d7dc0e268094b276b812c84fa41795892b78e2be94fd0cd303370e'::text)
                     Filter: (event_offset <= '92862790'::text)
               ->  Index Scan using participant_events_consuming_exercise_contract_id_idx on participant_events_consuming_exercise  (cost=0.00..2.22 rows=1 width=232) (actual time=0.014..0.014 rows=1 loops=1)
                     Index Cond: (contract_id = '004b4007da46aa40f0027345ba00bd542f581938b85fc584bb0be1e6c405847368ca011220116fd8c958d7dc0e268094b276b812c84fa41795892b78e2be94fd0cd303370e'::text)
                     Filter: (event_offset <= '92862790'::text)
 Planning Time: 0.521 ms
 Execution Time: 0.107 ms
(13 rows)
```

And this is the bulk query with 10 contracts:
```
 Append  (cost=44.63..45.19 rows=11 width=122) (actual time=0.196..0.342 rows=10 loops=1)
   CTE joined_rows
     ->  Nested Loop Left Join  (cost=11.08..44.63 rows=10 width=625) (actual time=0.108..0.265 rows=10 loops=1)
           ->  Bitmap Heap Scan on participant_events_create c  (cost=11.08..22.33 rows=10 width=577) (actual time=0.088..0.136 rows=10 loops=1)
                 Recheck Cond: (contract_id = ANY ('{004b4007da46aa40f0027345ba00bd542f581938b85fc584bb0be1e6c405847368ca011220116fd8c958d7dc0e268094b276b812c84fa41795892b78e2be94fd0cd303370e,00fe4c795783dca2b0bd74c3f3f580c131b2d03939de2c0c39cc1f29bba9fcc7aaca01122013bb3717724e20ce84366be050c545e342aedab357b4cfb6829a66d042bab336,001d0ce61bd0557c1f9038c6747f6dd58e5ad9c861f5b78a994bfc4dbbea7b0985ca0112201fa1829b532c8537cfe2326851ff487653321903d61bc3787de0a9767c453d91,00ad705699f95220afe56bf5e9894a5de894aafed9623c1d2f91e4f814b297b00eca011220d47dbfb03b852e1f75c07f16a7fe85101e55559b53535d67e9f987ead9ba2291,0021b8fdef56b6bd09c6dbd4063b91ca537b86142adddc655b3b55e2b1f351f0f1ca0112209f871746061068b040cb6f221d27dd72bfaaea2aaccf58d603e2815b3bc55c6e,00ce8d0397d89f743e9c4618dcd6e813fc19feb7648195351f898fa67d39f70482ca011220498e768c9b75592cfac59900a61e73daf2d797021f17b3419444b2952aeee968,0023055bb27ecd5300bb13c3c549c00567726a18b8645d1c1426f17d3eb7f0fe19ca011220aa6e599672d90589b27b69fcf70195d250ca986e7a9f0979f6d09942e694fe88,00d6a82f3ac8ff863d0eeb0c5699fb590f1c932705f86629153dd519b1d3ea07e3ca011220a0854f30fac6d84c208cc4905bf6c42ab2cf6e88a7ad4640925dcd99fc928507,0047b4ec4c89745f108444f64b7cbb73bf1f37eb6f2910236484a151cad7604bacca011220a25152eaa859cfea002e112600c97d93fc5ee828a0917a5218f0e9c319561907,00388d87b801425dc5d1b29215375d80d3e19af969c36835ad24a0e4ed10b2dc3eca011220ea8f008d2b28c3fb1dc60b95bd0aa46afbf2bb0509f7a5e29639d5a767c94bd0}'::text[]))
                 Rows Removed by Index Recheck: 1
                 Filter: (event_offset < '92862790'::text)
                 Heap Blocks: exact=10
                 ->  Bitmap Index Scan on participant_events_create_contract_id_idx  (cost=0.00..11.07 rows=10 width=0) (actual time=0.072..0.072 rows=11 loops=1)
                       Index Cond: (contract_id = ANY ('{004b4007da46aa40f0027345ba00bd542f581938b85fc584bb0be1e6c405847368ca011220116fd8c958d7dc0e268094b276b812c84fa41795892b78e2be94fd0cd303370e,00fe4c795783dca2b0bd74c3f3f580c131b2d03939de2c0c39cc1f29bba9fcc7aaca01122013bb3717724e20ce84366be050c545e342aedab357b4cfb6829a66d042bab336,001d0ce61bd0557c1f9038c6747f6dd58e5ad9c861f5b78a994bfc4dbbea7b0985ca0112201fa1829b532c8537cfe2326851ff487653321903d61bc3787de0a9767c453d91,00ad705699f95220afe56bf5e9894a5de894aafed9623c1d2f91e4f814b297b00eca011220d47dbfb03b852e1f75c07f16a7fe85101e55559b53535d67e9f987ead9ba2291,0021b8fdef56b6bd09c6dbd4063b91ca537b86142adddc655b3b55e2b1f351f0f1ca0112209f871746061068b040cb6f221d27dd72bfaaea2aaccf58d603e2815b3bc55c6e,00ce8d0397d89f743e9c4618dcd6e813fc19feb7648195351f898fa67d39f70482ca011220498e768c9b75592cfac59900a61e73daf2d797021f17b3419444b2952aeee968,0023055bb27ecd5300bb13c3c549c00567726a18b8645d1c1426f17d3eb7f0fe19ca011220aa6e599672d90589b27b69fcf70195d250ca986e7a9f0979f6d09942e694fe88,00d6a82f3ac8ff863d0eeb0c5699fb590f1c932705f86629153dd519b1d3ea07e3ca011220a0854f30fac6d84c208cc4905bf6c42ab2cf6e88a7ad4640925dcd99fc928507,0047b4ec4c89745f108444f64b7cbb73bf1f37eb6f2910236484a151cad7604bacca011220a25152eaa859cfea002e112600c97d93fc5ee828a0917a5218f0e9c319561907,00388d87b801425dc5d1b29215375d80d3e19af969c36835ad24a0e4ed10b2dc3eca011220ea8f008d2b28c3fb1dc60b95bd0aa46afbf2bb0509f7a5e29639d5a767c94bd0}'::text[]))
           ->  Index Scan using participant_events_consuming_exercise_contract_id_idx on participant_events_consuming_exercise a  (cost=0.00..2.22 rows=1 width=190) (actual time=0.008..0.011 rows=1 loops=10)
                 Index Cond: (contract_id = c.contract_id)
                 Filter: (event_offset < '92862790'::text)
   ->  CTE Scan on joined_rows  (cost=0.00..0.20 rows=1 width=122) (actual time=0.195..0.325 rows=1 loops=1)
         Filter: (archive_event IS NULL)
         Rows Removed by Filter: 9
   ->  CTE Scan on joined_rows joined_rows_1  (cost=0.00..0.20 rows=10 width=122) (actual time=0.004..0.010 rows=9 loops=1)
         Filter: (archive_event IS NOT NULL)
         Rows Removed by Filter: 1
 Planning Time: 0.818 ms
 Execution Time: 0.458 ms
(21 rows)
```

Reply:
Promising plan :+1: